api/bootstrap/kubeadm/v1beta2/kubeadmconfig_types.go

```diff
@@ -166,29 +166,34 @@ func (c *KubeadmConfigSpec) Validate(pathPrefix *field.Path) field.ErrorList {
 		}
 	}
 
-	// Validate timeouts
-	// Note: When v1beta1 will be removed, we can drop this limitation.
-	tInit := "unset"
-	if c.InitConfiguration.Timeouts.ControlPlaneComponentHealthCheckSeconds != nil {
-		tInit = fmt.Sprintf("%d", *c.InitConfiguration.Timeouts.ControlPlaneComponentHealthCheckSeconds)
-	}
-	tJoin := "unset"
-	if c.JoinConfiguration.Timeouts.ControlPlaneComponentHealthCheckSeconds != nil {
-		tJoin = fmt.Sprintf("%d", *c.JoinConfiguration.Timeouts.ControlPlaneComponentHealthCheckSeconds)
-	}
-	if tInit != tJoin {
-		allErrs = append(allErrs,
-			field.Invalid(
-				pathPrefix.Child("initConfiguration", "timeouts", "controlPlaneComponentHealthCheckSeconds"),
-				tInit,
-				fmt.Sprintf("controlPlaneComponentHealthCheckSeconds must be set to the same value both in initConfiguration.timeouts (%s) and in joinConfiguration.timeouts (%s)", tInit, tJoin),
-			),
-			field.Invalid(
-				pathPrefix.Child("joinConfiguration", "timeouts", "controlPlaneComponentHealthCheckSeconds"),
-				tJoin,
-				fmt.Sprintf("controlPlaneComponentHealthCheckSeconds must be set to the same value both in initConfiguration.timeouts (%s) and in joinConfiguration.timeouts (%s)", tInit, tJoin),
-			),
-		)
+	// Only ensure ControlPlaneComponentHealthCheckSeconds fields are equal for KubeadmControlPlane and KubeadmControlPlaneTemplate.
+	// In KubeadmConfig objects usually only one of InitConfiguration or JoinConfiguration is defined as a Machine uses
+	// either kubeadm init or kubeadm join, but not both.
+	if isKCP {
```

Comment:
We can't enforce this for KubeadmConfig / KubeadmConfigTemplate.

KubeadmConfig will only have InitConfiguration or JoinConfiguration
KubeadmConfigTemplate will only have JoinConfiguration (it's only for worker nodes and they only use kubeadm join)

Anyway, this whole validation block will be removed when we drop v1beta1